cli/cmd/top.go

```diff
@@ -0,0 +1,393 @@
+			}
+
+			req, err := util.BuildTapByResourceRequest(requestParams)
+			if err != nil {
+				return err
+			}
+
+			return getTrafficByResourceFromAPI(os.Stdout, validatedPublicAPIClient(), req)
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace,
+		"Namespace of the specified resource")
+	cmd.PersistentFlags().StringVar(&options.toResource, "to", options.toResource,
+		"Display requests to this resource")
+	cmd.PersistentFlags().StringVar(&options.toNamespace, "to-namespace", options.toNamespace,
+		"Sets the namespace used to lookup the \"--to\" resource; by default the current \"--namespace\" is used")
+	cmd.PersistentFlags().Float32Var(&options.maxRps, "max-rps", options.maxRps,
+		"Maximum requests per second to tap.")
+	cmd.PersistentFlags().StringVar(&options.scheme, "scheme", options.scheme,
+		"Display requests with this scheme")
+	cmd.PersistentFlags().StringVar(&options.method, "method", options.method,
+		"Display requests with this HTTP method")
+	cmd.PersistentFlags().StringVar(&options.authority, "authority", options.authority,
+		"Display requests with this :authority")
+	cmd.PersistentFlags().StringVar(&options.path, "path", options.path,
+		"Display requests with paths that start with this prefix")
+
+	return cmd
+}
+
+func getTrafficByResourceFromAPI(w io.Writer, client pb.ApiClient, req *pb.TapByResourceRequest) error {
+
+	rsp, err := client.TapByResource(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	err = termbox.Init()
+	if err != nil {
+		return err
+	}
+	defer termbox.Close()
+
+	requestCh := make(chan topRequest, 100)
+	done := make(chan struct{})
+
+	go recvEvents(rsp, requestCh, done)
+	go pollInput(done)
+
+	renderTable(requestCh, done)
+
+	return nil
+}
+
+func recvEvents(tapClient pb.Api_TapByResourceClient, requestCh chan<- topRequest, done chan<- struct{}) {
+	outstandingRequests := make(map[topRequestID]topRequest)
+	for {
+		event, err := tapClient.Recv()
+		if err == io.EOF {
+			fmt.Println("Tap stream terminated")
+			close(done)
+			return
+		}
+		if err != nil {
+			fmt.Println(err.Error())
+			close(done)
+			return
+		}
+
+		switch ev := event.GetHttp().GetEvent().(type) {
+		case *pb.TapEvent_Http_RequestInit_:
+			id := topRequestID{
+				addr.PublicAddressToString(event.GetSource()),
+				addr.PublicAddressToString(event.GetDestination()),
+				ev.RequestInit.GetId().Stream,
+			}
+
+			outstandingRequests[id] = topRequest{
+				event:   event,
+				reqInit: ev.RequestInit,
+			}
+
+		case *pb.TapEvent_Http_ResponseInit_:
+			id := topRequestID{
+				addr.PublicAddressToString(event.GetSource()),
+				addr.PublicAddressToString(event.GetDestination()),
+				ev.ResponseInit.GetId().Stream,
+			}
+			if req, ok := outstandingRequests[id]; ok {
+				req.rspInit = ev.ResponseInit
+				outstandingRequests[id] = req
+			} else {
+				log.Warnf("Got ResponseInit for unknown stream: %s->%s(%d)", id.src, id.dst, id.stream)
```

Comment:
TIOLI you could add a `String()` func to your `topRequestID` struct that will do the string formatting automatically. Something like:

```
func (id topRequestID) String() string {
	return fmt.Sprintf("%s->%s(%d)", id.src, id.dst, id.stream)
}
```

And then here and in the other log statement below, you could just let the id be coerced to a string, which will invoke that function:

```
log.Warnf("Got ResponseInit for unknown stream: %s", id)
```